dev/AnimatedIcon/AnimatedIcon.cpp

```diff
@@ -0,0 +1,516 @@
+                if (found)
+                {
+                    m_progressPropertySet.InsertScalar(L"Progress", value);
+                }
+                else
+                {
+                    // We also support setting the state proprety to a float value, which instructs the animated icon
+                    // to animated the Progress property to the provided value.
+                    try
+                    {
+                        auto const stateAsFloat = std::stof(toState.data());
+                        PlaySegment(NAN, stateAsFloat, playbackMultiplier);
+                        m_lastAnimationSegmentStart = L"";
+                        m_lastAnimationSegmentEnd = toState;
+                    }
+                    catch (...)
+                    {
+                        // None of our attempt to find an animation to play or frame to show have worked, so just cut
+                        // to frame 0.
+                        m_progressPropertySet.InsertScalar(L"Progress", 0.0);
+                        m_lastAnimationSegmentStart = L"";
+                        m_lastAnimationSegmentEnd = L"0.0";
+                    }
+                }
+            }
+            m_lastAnimationSegment = transitionName;
+            AnimatedIconTestHooks::NotifyLastAnimationSegmentChanged(*this);
+        }
+    }
+}
+
+void AnimatedIcon::PlaySegment(float from, float to, float playbackMultiplier)
+{
+    auto const segmentLength = [from, to, previousSegmentLength = m_previousSegmentLength]()
+    {
+        if (std::isnan(from))
+        {
+            return previousSegmentLength;
+        }
+        return std::abs(to - from);
+    }();
+
+    m_previousSegmentLength = segmentLength;
+    auto const duration = m_animatedVisual ?
+        std::chrono::duration_cast<winrt::TimeSpan>(m_animatedVisual.get().Duration() * segmentLength * (1.0 / playbackMultiplier) * m_durationMultiplier) :
+        winrt::TimeSpan::zero();
+    // If the duration is really short (< 20ms) don't bother trying to animate.
+    if (duration < winrt::TimeSpan{ 20ms })
+    {
+        m_progressPropertySet.InsertScalar(L"Progress", to);
+        OnAnimationCompleted(nullptr, nullptr);
+    }
+    else
+    {
+        auto compositor = m_progressPropertySet.Compositor();
+        auto animation = compositor.CreateScalarKeyFrameAnimation();
+        animation.Duration(duration);
+        auto linearEasing = compositor.CreateLinearEasingFunction();
+
+        // Play from fromProgress.
+        if (!std::isnan(from))
+        {
+            //m_progressPropertySet.InsertScalar(L"Progress", from);
+            animation.InsertKeyFrame(0, from);
+        }
+
+        // Play to toProgress
+        animation.InsertKeyFrame(1, to, linearEasing);
+
+        animation.IterationBehavior(winrt::AnimationIterationBehavior::Count);
+        animation.IterationCount(1);
+
+        if (m_batch)
+        {
+            m_batch.Completed(m_batchCompletedToken);
```

Comment:
I don't see this event getting detached in the destructor, which is another reason to do the revoker pattern.